docs/proposals/2170-kubeflow-training-v2/README.md

```diff
@@ -0,0 +1,1362 @@
+public repository that users can apply to their clusters.
+
+Platform or ML engineers can tweak existing blueprint, based on their requirements. For example,
+using custom configurations.
+
+The Kubeflow Training Operator can maintain more Training Runtimes when the community is ready to
+support them. For example, runtimes for [Jax](https://jax.readthedocs.io/en/latest/index.html) or
+[MLX](https://ml-explore.github.io/mlx/build/html/index.html). Initially, we will support: PyTorch,
+MPI, TensorFlow, XGBoost, and PaddlePaddle.
+
+The `TrainingRuntime` is immutable, and so to make a change, a new version of the `TrainingRuntime`
+must be created and then changing the `TranJob` to point to the new version.
+This provides control as to how changes to runtimes propagate to existing training jobs.
+For example, when training is running for a long time (e.g. 1-2 months).
+
+In the future implementation, we will introduce a revision control mechanism similar to
+[Kubernetes Deployment](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#updating-a-deployment)
+to control versions of `TrainingRuntime` and enable rolling updates.
+
+We are going to create two CRDs: `TrainingRuntime` and `ClusterTrainingRuntime`. These runtimes have
+exactly the same APIs, but the first one is the namespace-scoped, the second is the cluster-scoped.
+If `trainingRuntimeRef` in `TrainJob` has the namespace, controller will use the `TrainingRuntime`,
+otherwise it will use the `ClusterTrainingRuntime`.
+
+```golang
+type TrainingRuntime struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    // Framework specific parameters.
+    MLSpec *MLSpec `json:"mlSpec,omitempty"`
+
+    // Number of nodes to execute training.
+    NumNodes int `json:"numNodes,omitempty"`
+
+    // JobSet spec.
+    JobSetSpec *batchv1.JobSetSpec `json:",inline"`
+
+    // For gang-scheduling using volcano or scheduler plugins, supported for all frameworks.
+    GangScheduler *kubeflowv1.GangScheduler `json:"gangScheduler,omitempty"`
+}
+
+// One of the specs can be selected.
+type MLSpec struct {
+
+    // Custom Spec for Torch
+    TorchSpec *TorchSpec `json:"torchSpec,omitempty"`
+
+    // Custom Spec for MPI
+    MPISpec *MPISpec `json:"mpiSpec,omitempty"`
+}
+```
+
+### The Gang Scheduler API
+
+Gang scheduler plugin is used to create appropriate `PodGroup` for Volcano or scheduler plugins.
+
+```golang
+type GangScheduler struct {
+    // Plugin for gang scheduling.
+    Plugin *GangSchedulerPlugin `json:plugin,omitempty"`
+
+    // Time threshold to schedule PodGroup for gang scheduling.
+    ScheduleTimeoutSeconds string `json:scheduleTimeoutSeconds,omitempty"`
+}
+
+type GangSchedulerPlugin string
+
+const (
+    GangSchedulerPluginVolcano GangSchedulerPlugin = "volcano"
+    GangSchedulerPluginSP      GangSchedulerPlugin = "scheduler-plugins"
```

Comment:
SP == service provider?

Reply:
It is when Scheduler Plugins are used for Gang Scheduling: https://github.com/kubernetes-sigs/scheduler-plugins?tab=readme-ov-file
Any thoughts on how we can make this name more clear ?

Reply:
What does it explicitly read to? 

GangSchedulerPluginSP == GangSchedulerPlugin Scheduler Plugin?

Reply:
Yes, it is for Scheduler Plugins.

Reply:
Perhaps just remove “SP” then, because it’s redundant - or add “Name” to one of the variables if needed.